File: cpp/tree/binary_tree_right_side_view.cpp
// 二叉树的右视图

#include "vector"
#include "stack"
#include "queue"

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() {}
    TreeNode(int val) : val(val) {}
    TreeNode(int val, TreeNode *left, TreeNode *right) : val(val), left(left), right(right) {}
};

class Solution {
public:

    // 广度优先搜索
    vector<int> rightSideView1(TreeNode* root) {
        vector<int> res;
        if (!root) {
            return res;
        }

        queue<TreeNode*> que;
        que.push(root);
        while (!que.empty()) {
            int size = que.size();
            while (size-- > 0) {
                TreeNode* node = que.front();
                que.pop();
                if (size == 0) {
                    res.push_back(node->val);
                }
                if (node->left) {
                    que.push(node->left);
                }
                if (node->right) {
                    que.push(node->right);
                }
            }
        }
        return res;
    }

    // 深度优先
    vector<int> rightSideView2(TreeNode* root) {
        vector<int> res;
        helper(root, res, 0);
        return res;
    }

    void helper(TreeNode* root, vector<int> &res, int level) {
        if (!root) {
            return;
        }
        if (level == res.size()) {
            res.push_back(root->val);
        }
        helper(root->right, res, level+1);
        helper(root->left, res, level+1);
    }
};

File: go/tree/leetcode/binary_tree_right_side_view.go
package leetcode

// 二叉树的右视图

// 广度优先
func rightSideView1(root *TreeNode) []int {
	var res []int
	if root == nil {
		return res
	}

	queue := []*TreeNode{root}
	for len(queue) > 0 {
		size := len(queue)
		for size > 0 {
			size--
			node := queue[0]
			queue = queue[1:]
			if size == 0 {
				res = append(res, node.Val)
			}
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}
	}
	return res
}

// 深度优先
func rightSideView2(root *TreeNode) []int {
	var res []int
	rightSideViewHelper(root, &res, 0)
	return res
}

func rightSideViewHelper(root *TreeNode, res *[]int, level int) {
	if root == nil {
		return
	}
	if level == len(*res) {
		*res = append(*res, root.Val)
	}
	rightSideViewHelper(root.Right, res, level+1)
	rightSideViewHelper(root.Left, res, level+1)
}
